File: java-vfs/src/main/java/emg/java/vfs/commands/CreateEntry.java
package emg.java.vfs.commands;

import emg.java.vfs.files.DirEntry;
import emg.java.vfs.files.Directory;
import emg.java.vfs.filesystem.State;

import java.util.List;

import static emg.java.vfs.extensions.ListExtensions.head;
import static emg.java.vfs.extensions.ListExtensions.tail;

public abstract class CreateEntry extends Command {
    private final String name;

    public CreateEntry(String name) {
        this.name = name;
    }

    @Override
    public State apply(State state) {
        var wd = state.wd();
        if (wd.hasEntry(name)) {
            return state.setMessage(String.format("Directory %s already exists", name));
        } else if (name.contains(Directory.SEPARATOR)) {
            return state.setMessage(String.format("%s must not contain separators", name));
        } else if (checkIllegal(name)) {
            return state.setMessage(String.format("%s is illegal entry name", name));
        } else {
            return doCreateEntry(state, name);
        }
    }

    private State doCreateEntry(State state, String name) {
        var wd = state.wd();
        //1 all directories in path
        var allDirsInPath = wd.getAllFoldersInPath();
        //2 create new directory in wd
        var newDirectory = createSpecificEntry(state, name);
        //3 update the directory structure
        var newRoot = updateStructure(state.root(), allDirsInPath, newDirectory);
        //4 find the new WD
        var newWD = newRoot.findDesendant(allDirsInPath);

        return State.apply(newRoot, newWD);
    }

    public abstract DirEntry createSpecificEntry(State state, String entryName);

    private Directory updateStructure(Directory currentDirectory, List<String> path, DirEntry newEntry) {
        if (path.isEmpty()) {
            return currentDirectory.addEntry(newEntry);
        } else {
            var oldEntry = currentDirectory.findEntry(head(path));
            return currentDirectory.replaceEntry(oldEntry.name(), updateStructure(oldEntry.asDirectory(),
                tail(path), newEntry));
        }
    }

    private boolean checkIllegal(String name) {
        return name.contains(".");
    }
}

File: kotlin-vfs/src/main/kotlin/emg/kotlin/vfs/commands/CreateEntry.kt
package emg.kotlin.vfs.commands

import emg.kotlin.vfs.extensions.head
import emg.kotlin.vfs.extensions.tail
import emg.kotlin.vfs.files.DirEntry
import emg.kotlin.vfs.files.Directory
import emg.kotlin.vfs.filesystem.State

abstract class CreateEntry constructor(private val name: String) : Command() {
    override fun apply(state: State): State {
        val wd = state.wd
        return when {
            wd.hasEntry(name) -> {
                state.setMessage("Directory $name already exists")
            }
            name.contains(Directory.SEPARATOR) -> {
                state.setMessage("$name must not contain separators")
            }
            checkIllegal(name) -> {
                state.setMessage("$name is illegal entry name")
            }
            else -> {
                doCreateEntry(state, name)
            }
        }
    }

    private fun doCreateEntry(state: State, name: String): State {
        val wd = state.wd
        //1 all directories in path
        val allDirsInPath = wd.getAllFoldersInPath()
        //2 create new directory in wd
        val newDirectory = createSpecificEntry(state, name)
        //3 update the directory structure
        val newRoot = updateStructure(state.root, allDirsInPath, newDirectory)
        //4 find the new WD
        val newWD = newRoot.findDesendant(allDirsInPath)

        return State.apply(newRoot, newWD)
    }

    private fun updateStructure(currentDirectory: Directory, path: List<String>, newEntry: DirEntry): Directory {
        return if (path.isEmpty()) currentDirectory.addEntry(newEntry)
        else {
            val oldEntry = currentDirectory.findEntry(path.head())!!
            currentDirectory.replaceEntry(oldEntry.name, updateStructure(oldEntry.asDirectory(),
                    path.tail(), newEntry))
        }
    }

    private fun checkIllegal(name: String): Boolean {
        return name.contains(".")
    }

    abstract fun createSpecificEntry(state: State, entryName: String): DirEntry

}